File: last-round/98.验证二叉搜索树.js
/*
 * @lc app=leetcode.cn id=98 lang=javascript
 *
 * [98] 验证二叉搜索树
 *
 * https://leetcode-cn.com/problems/validate-binary-search-tree/description/
 *
 * algorithms
 * Medium (34.19%)
 * Likes:    1056
 * Dislikes: 0
 * Total Accepted:    269.2K
 * Total Submissions: 787.1K
 * Testcase Example:  '[2,1,3]'
 *
 * 给定一个二叉树，判断其是否是一个有效的二叉搜索树。
 * 
 * 假设一个二叉搜索树具有如下特征：
 * 
 * 
 * 节点的左子树只包含小于当前节点的数。
 * 节点的右子树只包含大于当前节点的数。
 * 所有左子树和右子树自身必须也是二叉搜索树。
 * 
 * 
 * 示例 1:
 * 
 * 输入:
 * ⁠   2
 * ⁠  / \
 * ⁠ 1   3
 * 输出: true
 * 
 * 
 * 示例 2:
 * 
 * 输入:
 * ⁠   5
 * ⁠  / \
 * ⁠ 1   4
 * / \
 * 3   6
 * 输出: false
 * 解释: 输入为: [5,1,4,null,null,3,6]。
 * 根节点的值为 5 ，但是其右子节点值为 4 。
 * 
 * 
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function(root) { 
  const recurse = (root) => {
    if (!root) return

    const left = recurse(root.left)
    const right = recurse(root.right)

    let isValid = true
    if (left) isValid = isValid && left[0] && left[2] < root.val
    if (right) isValid = isValid && right[0] && right[1] > root.val

    return [
      isValid,
      Math.min(left ? left[1] : Infinity, right ? right[1] : Infinity, root.val),
      Math.max(right ? right[2] : -Infinity, left ? left[2] : -Infinity, root.val),
    ]
  }

  // console.log(recurse(root))
  return recurse(root)[0]
};
// @lc code=end

File: second-round/98.验证二叉搜索树.py
#
# @lc app=leetcode.cn id=98 lang=python3
#
# [98] 验证二叉搜索树
#
# https://leetcode-cn.com/problems/validate-binary-search-tree/description/
#
# algorithms
# Medium (35.37%)
# Likes:    1387
# Dislikes: 0
# Total Accepted:    407.4K
# Total Submissions: 1.2M
# Testcase Example:  '[2,1,3]'
#
# 给你一个二叉树的根节点 root ，判断其是否是一个有效的二叉搜索树。
# 
# 有效 二叉搜索树定义如下：
# 
# 
# 节点的左子树只包含 小于 当前节点的数。
# 节点的右子树只包含 大于 当前节点的数。
# 所有左子树和右子树自身必须也是二叉搜索树。
# 
# 
# 
# 
# 示例 1：
# 
# 
# 输入：root = [2,1,3]
# 输出：true
# 
# 
# 示例 2：
# 
# 
# 输入：root = [5,1,4,null,null,3,6]
# 输出：false
# 解释：根节点的值是 5 ，但是右子节点的值是 4 。
# 
# 
# 
# 
# 提示：
# 
# 
# 树中节点数目范围在[1, 10^4] 内
# -2^31 <= Node.val <= 2^31 - 1
# 
# 
#


# Definition for a binary tree node.
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# @lc code=start
class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        queue = deque([(root, float('-inf'), float('inf'))])

        while queue:
            curr, min_val, max_val = queue.popleft()

            if not min_val < curr.val < max_val:
                return False

            if curr.left:
                queue.append((curr.left, min_val, min(max_val, curr.val)))
            if curr.right:
                queue.append((curr.right, max(min_val, curr.val), max_val))

        return True
# @lc code=end
